doc/user-manual/language/type-based-termination-checking.rst

```diff
@@ -0,0 +1,317 @@
+
+For example, consider a coinductive function ``zipWith``:
+
+::
+
+    zipWith : (Nat → Nat → Nat) → Stream → Stream → Stream
+    zipWith f s1 s2 .head = f (s1 .head) (s2 .head)
+    zipWith f s1 s2 .tail = zipWith f (s1 .tail) (s2 .tail)
+
+Here, the depth of the returned ``Stream`` is the same as the requested depth of incoming ``s1`` and ``s2``. The type-based termination checker recognizes this, concluding that all three ``s1``, ``s2``, and the returned stream share the same size variable.
+
+Given size-preserving ``zipWith``, Agda is able to define an infinite stream of Fibonacci numbers:
+
+::
+
+   fib : Stream
+   fib .head = zero
+   fib .tail .head = suc zero
+   fib .tail .tail = zipWith plus fib (fib .tail)
+
+This function passes termination checking. We shall explain the logic of Agda for the third clause.
+
+Following our intuition with coinductive functions, the are three depth parameters ``k < m < n``, where the outer stream is of depth ``n``, and to pass checking the third clause should return the stream of depth at least ``k``. If the first inner ``fib`` is used with the depth ``k``, and the second ``fib`` is used with the depth ``m`` (note, that the smallest available depth for ``fib .tail`` is ``k``, hence ``fib`` must use something bigger, which is ``m``), then the size-preserving ``zipWith`` returns a stream of size ``k``, which is indeed what is required from it. Now we see that both recursive calls to ``fib`` are performed with depths ``k`` and ``m``, which are smaller than ``n``. Agda concludes that this function is terminating.
+
+Size preservation is tightly coupled with polarities. Given a function signature, all occurrences of *negative inductive* and *positive coinductive* variables are considered as input, and they serve as possible candidates for size preservation analysis. On the other hand, all *positive inductive* and *negative coinductive* variables are considered as output, and a function signature may be size-preserving precisely in them. For example, consider the following definition:
+
+::
+
+    foo : {A : Set} → (Nat → A) → Nat → A
+    foo f x = f x
+
+
+
+Here, the first ``Nat`` in ``foo`` is in a doubly-negative position, which means that the position is positive, and ``foo`` can be size-preserving in the first ``Nat``. From the definition we see that it is indeed the case. One application of this fact is that the following function passes termination check:
+
+::
+
+    bar : Nat → Nat
+    bar zero = zero
+    bar (suc n) = foo bar n
+
+As a final note, we address performance considerations. Currently, size-preservation analysis is the slowest part of the type-based termination checker. If you suspect that it causes a slowdown, you can specify ``--no-size-preservation``, disabling the analysis while retaining the rest of the type-based termination checker. Nevertheless, there are plans to improve its performance.
```

Comment:
I don't think the user manual is the right place to document future changes or enhancements to Agda, so I would leave out the final sentence.